pkg/webhooks/webhook.go

```diff
@@ -0,0 +1,300 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/mattbaird/jsonpatch"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/api/extensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientadmissionregistrationv1beta1 "k8s.io/client-go/kubernetes/typed/admissionregistration/v1beta1"
+)
+
+const (
+	secretServerKey  = "server-key.pem"
+	secretServerCert = "server-cert.pem"
+	secretCACert     = "ca-cert.pem"
+)
+
+var deploymentKind = v1beta1.SchemeGroupVersion.WithKind("Deployment")
+
+// Run implements the admission controller run loop.
+func (ac *AdmissionController) Run(stop <-chan struct{}) error {
+	tlsConfig, caCert, err := BuildCerts(ac.Client, &ac.Options)
+	if err != nil {
+		glog.Infof("could not configure admission webhook certs: %v", err)
+		return err
+	}
+
+	server := &http.Server{
+		Handler:   ac,
+		Addr:      fmt.Sprintf(":%v", ac.Options.Port),
+		TLSConfig: tlsConfig,
+	}
+
+	glog.Info("Found certificates for webhook...")
+	if ac.Options.RegistrationDelay != 0 {
+		glog.Infof("Delaying admission webhook registration for %v", ac.Options.RegistrationDelay)
+	}
+
+	select {
+	case <-time.After(ac.Options.RegistrationDelay):
+		cl := ac.Client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations()
+		if err := ac.Register(cl, caCert); err != nil {
```

Comment:
Should these functions be public, or are they only called from here?  We should start with the minimal visibility.